From e2aaef4be71af2741ee8db1a1399328ee193528f Mon Sep 17 00:00:00 2001
From: Lawrence Rust <lvr@softsystem.co.uk>
Date: Mon, 5 Aug 2013 14:44:11 +0100
Subject: [PATCH 026/227] RingBuffer: Prevent a race condition during
 RingBuffer destruction

If the readahead thread is running while deleting a RingBuffer
derived class then the thread can call the virtual safe_read method
which results in undefined behaviour if the vtable has been reset
as a normal part of the derived class' dtor.

Signed-off-by: Lawrence Rust <lvr@softsystem.co.uk>
---
 mythtv/libs/libmythtv/Bluray/bdringbuffer.cpp |    2 ++
 mythtv/libs/libmythtv/DVD/dvdringbuffer.cpp   |    2 ++
 mythtv/libs/libmythtv/fileringbuffer.cpp      |    2 ++
 3 files changed, 6 insertions(+)

diff --git a/mythtv/libs/libmythtv/Bluray/bdringbuffer.cpp b/mythtv/libs/libmythtv/Bluray/bdringbuffer.cpp
index 5469f54..695b50b 100644
--- a/mythtv/libs/libmythtv/Bluray/bdringbuffer.cpp
+++ b/mythtv/libs/libmythtv/Bluray/bdringbuffer.cpp
@@ -52,6 +52,8 @@ BDRingBuffer::BDRingBuffer(const QString &lfilename)
 
 BDRingBuffer::~BDRingBuffer()
 {
+    KillReadAheadThread();
+
     close();
 }
 
diff --git a/mythtv/libs/libmythtv/DVD/dvdringbuffer.cpp b/mythtv/libs/libmythtv/DVD/dvdringbuffer.cpp
index 69f78b1..d2e3e25 100644
--- a/mythtv/libs/libmythtv/DVD/dvdringbuffer.cpp
+++ b/mythtv/libs/libmythtv/DVD/dvdringbuffer.cpp
@@ -136,6 +136,8 @@ DVDRingBuffer::DVDRingBuffer(const QString &lfilename) :
 
 DVDRingBuffer::~DVDRingBuffer()
 {
+    KillReadAheadThread();
+
     CloseDVD();
     m_menuBtnLock.lock();
     ClearMenuSPUParameters();
diff --git a/mythtv/libs/libmythtv/fileringbuffer.cpp b/mythtv/libs/libmythtv/fileringbuffer.cpp
index 07aad39..1deb505 100644
--- a/mythtv/libs/libmythtv/fileringbuffer.cpp
+++ b/mythtv/libs/libmythtv/fileringbuffer.cpp
@@ -86,6 +86,8 @@ FileRingBuffer::FileRingBuffer(const QString &lfilename,
 
 FileRingBuffer::~FileRingBuffer()
 {
+    KillReadAheadThread();
+
     rwlock.lockForWrite();
 
     if (remotefile)
-- 
1.7.9.5

